File: Java/src/main/java/cc/Trie.java
package cc;

public class Trie {
	private TrieNode root;

	public Trie() {
		root = new TrieNode();
	}

	public void insert(String word) {
		TrieNode current = root;

		for (char l : word.toCharArray())
			current = current.getChildren().computeIfAbsent(l, c -> new TrieNode());

		current.setEndOfWord(true);
	}

	public boolean search(String word) {
		TrieNode current = root;

		for (int i = 0; i < word.length(); i++) {
			char ch = word.charAt(i);
			TrieNode node = current.getChildren().get(ch);

			if (node == null) return false;

			current = node;
		}

		return current.isEndOfWord();
	}
}

File: Python/data_structures/trie.py
class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.is_end_of_word = False


class Trie:
    def __init__(self):
        self.root = {}

    # noinspection PyMethodMayBeStatic
    def get_node(self):
        return TrieNode()

    # noinspection PyMethodMayBeStatic
    def char_to_index(self, char):
        return ord(char) - ord('a')

    def insert(self, word) -> None:
        start = self.root
        for i in word:
            if i not in start:
                start[i] = {}
            start = start[i]
        start['$'] = True

    def search(self, word):
        start = self.root
        for i in word:
            if i not in start:
                return False
            start = start[i]
        return '$' in start
